File: ClinicaVeterinaria/logica/GestorClinica.cs
using System;
using System.Collections.Generic;
using System.Linq;

namespace ClinicaVeterinaria.logica
{
    public class GestorClinica
    {
        public List<Veterinario> lstVeterinarios { get; set; }
        public List<Cliente> lstClientes { get; set; }
        public List<Mascota> lstMascotas { get; set; }
        public List<Ticket> lstTickets { get; set; }
        private FlujoTicket objFlujo;
        private int intSiguienteNumero;

        public GestorClinica()
        {
            lstVeterinarios = new List<Veterinario>();
            lstClientes = new List<Cliente>();
            lstMascotas = new List<Mascota>();
            lstTickets = new List<Ticket>();
            objFlujo = new FlujoTicket();
            intSiguienteNumero = 1;
        }

        public Ticket crearTicket(string strMotivo, string strSintomas, string strTipoConsulta, string strPrioridad, Mascota objMascota)
        {
            Ticket objTicket = new Ticket(intSiguienteNumero, strMotivo, strSintomas, strTipoConsulta, strPrioridad, objMascota);
            intSiguienteNumero++;
            lstTickets.Add(objTicket);
            try
            {
                asignarTicket(objTicket.intNumero);
            }
            catch (InvalidOperationException)
            {
                // El ticket queda Abierto si no hay veterinario disponible por el momento.
            }
            return objTicket;
        }

        public Ticket asignarTicket(int intNumero)
        {
            Ticket objTicket = buscarTicket(intNumero);
            Veterinario objVeterinario = asignarVeterinarioAutomatico(objTicket.strTipoConsulta);
            if (objVeterinario == null)
                throw new InvalidOperationException("No hay veterinarios disponibles para este tipo de consulta.");
            objTicket.asignar(objVeterinario);
            return objTicket;
        }

        public Ticket asignarTicket(int intNumero, string strIdVeterinario)
        {
            Ticket objTicket = buscarTicket(intNumero);
            Veterinario objVeterinario = lstVeterinarios.FirstOrDefault(v => v.strId == strIdVeterinario);
            if (objVeterinario == null)
                throw new KeyNotFoundException("Veterinario no encontrado.");
            objTicket.asignar(objVeterinario);
            return objTicket;
        }

        public Ticket registrarHallazgo(int intNumero, string strTipo, string strDescripcion, string strGravedad)
        {
            Ticket objTicket = buscarTicket(intNumero);
            objTicket.registrarHallazgo(strTipo, strDescripcion, strGravedad);
            bool blnGravedadAlta = strGravedad.Equals("Alta", StringComparison.OrdinalIgnoreCase)
                || strGravedad.Equals("Critica", StringComparison.OrdinalIgnoreCase);
            if (blnGravedadAlta && objFlujo.puedeCambiarEstado(objTicket.strEstado, "Escalado"))
                escalarTicket(intNumero, $"Hallazgo clinico de gravedad {strGravedad}");
            return objTicket;
        }

        public Ticket escalarTicket(int intNumero, string strMotivo)
        {
            Ticket objTicket = buscarTicket(intNumero);
            if (!objFlujo.puedeCambiarEstado(objTicket.strEstado, "Escalado"))
                throw new InvalidOperationException($"No se puede escalar un ticket en estado '{objTicket.strEstado}'.");
            Veterinario objSenior = lstVeterinarios
                .Where(v => v.strEspecialidad.Equals(objTicket.strTipoConsulta, StringComparison.OrdinalIgnoreCase)
                    || v.strEspecialidad.Equals("General", StringComparison.OrdinalIgnoreCase))
                .OrderByDescending(v => v.intNivelExperiencia)
                .FirstOrDefault();
            objTicket.escalar(strMotivo, objSenior);
            return objTicket;
        }

        public Ticket resolverTicket(int intNumero, string strDiagnostico, string strTratamiento)
        {
            Ticket objTicket = buscarTicket(intNumero);
            if (!objFlujo.puedeCambiarEstado(objTicket.strEstado, "Resuelto"))
                throw new InvalidOperationException($"No se puede resolver un ticket en estado '{objTicket.strEstado}'.");
            objTicket.resolver(strDiagnostico, strTratamiento);
            return objTicket;
        }

        public Ticket cerrarTicket(int intNumero)
        {
            Ticket objTicket = buscarTicket(intNumero);
            if (!objFlujo.puedeCambiarEstado(objTicket.strEstado, "Cerrado"))
                throw new InvalidOperationException($"No se puede cerrar un ticket en estado '{objTicket.strEstado}'.");
            objTicket.cerrar();
            return objTicket;
        }

        private Veterinario asignarVeterinarioAutomatico(string strTipoConsulta)
        {
            return lstVeterinarios
                .Where(v => v.puedeAtender(strTipoConsulta))
                .OrderBy(v => v.intCargaActual)
                .FirstOrDefault();
        }

        public Ticket buscarTicket(int intNumero)
        {
            Ticket objTicket = lstTickets.FirstOrDefault(t => t.intNumero == intNumero);
            if (objTicket == null)
                throw new KeyNotFoundException($"No existe un ticket con numero {intNumero}.");
            return objTicket;
        }

        public void mostrarTickets()
        {
            if (lstTickets.Count == 0)
            {
                Console.WriteLine("No hay tickets registrados.");
                return;
            }
            foreach (var objTicket in lstTickets)
                objTicket.mostrarResumen();
        }

        public void mostrarFlujo()
        {
            objFlujo.mostrarFlujo();
        }

        public void generarMetricas()
        {
            Console.WriteLine($"Total de tickets: {lstTickets.Count}");

            string[] arrEstados = { "Abierto", "Asignado", "Escalado", "Resuelto", "Cerrado" };
            Console.WriteLine("Por estado:");
            foreach (var strEstado in arrEstados)
                Console.WriteLine($"  {strEstado}: {lstTickets.Count(t => t.strEstado == strEstado)}");

            string[] arrPrioridades = { "Baja", "Media", "Alta", "Emergencia" };
            Console.WriteLine("Por prioridad:");
            foreach (var strPrioridad in arrPrioridades)
                Console.WriteLine($"  {strPrioridad}: {lstTickets.Count(t => t.strPrioridad.Equals(strPrioridad, StringComparison.OrdinalIgnoreCase))}");

            Console.WriteLine($"Tickets escalados: {lstTickets.Count(t => t.blnEscalado)}");

            Console.WriteLine("Carga por veterinario:");
            foreach (var objVeterinario in lstVeterinarios)
                Console.WriteLine($"  {objVeterinario.strNombre} ({objVeterinario.strEspecialidad}): {objVeterinario.intCargaActual}/{objVeterinario.intCapacidadMaxima}");
        }
    }
}

File: BlueJ_Proyecto/GestorClinica.java
import java.util.ArrayList;
import java.util.NoSuchElementException;

/**
 * Coordina veterinarios, clientes, mascotas y tickets de la clinica.
 */
public class GestorClinica
{
    private ArrayList<Veterinario> veterinarios;
    private ArrayList<Cliente> clientes;
    private ArrayList<Mascota> mascotas;
    private ArrayList<Ticket> tickets;
    private FlujoTicket flujo;
    private int siguienteNumero;

    public GestorClinica()
    {
        veterinarios = new ArrayList<>();
        clientes = new ArrayList<>();
        mascotas = new ArrayList<>();
        tickets = new ArrayList<>();
        flujo = new FlujoTicket();
        siguienteNumero = 1;
    }

    public Ticket crearTicket(String motivo, String sintomas, String tipoConsulta, String prioridad, Mascota mascota)
    {
        Ticket ticket = new Ticket(siguienteNumero, motivo, sintomas, tipoConsulta, prioridad, mascota);
        siguienteNumero++;
        tickets.add(ticket);
        try
        {
            asignarTicket(ticket.getNumero());
        }
        catch (IllegalStateException ex)
        {
            // El ticket queda Abierto si no hay veterinario disponible por el momento.
        }
        return ticket;
    }

    public Ticket asignarTicket(int numero)
    {
        Ticket ticket = buscarTicket(numero);
        Veterinario veterinario = asignarVeterinarioAutomatico(ticket.getTipoConsulta());
        if (veterinario == null)
        {
            throw new IllegalStateException("No hay veterinarios disponibles para este tipo de consulta.");
        }
        ticket.asignar(veterinario);
        return ticket;
    }

    public Ticket asignarTicket(int numero, String idVeterinario)
    {
        Ticket ticket = buscarTicket(numero);
        Veterinario veterinario = null;
        for (Veterinario v : veterinarios)
        {
            if (v.getId().equals(idVeterinario))
            {
                veterinario = v;
                break;
            }
        }
        if (veterinario == null)
        {
            throw new NoSuchElementException("Veterinario no encontrado.");
        }
        ticket.asignar(veterinario);
        return ticket;
    }

    public Ticket registrarHallazgo(int numero, String tipo, String descripcion, String gravedad)
    {
        Ticket ticket = buscarTicket(numero);
        ticket.registrarHallazgo(tipo, descripcion, gravedad);
        boolean gravedadAlta = gravedad.equalsIgnoreCase("Alta") || gravedad.equalsIgnoreCase("Critica");
        if (gravedadAlta && flujo.puedeCambiarEstado(ticket.getEstado(), "Escalado"))
        {
            escalarTicket(numero, "Hallazgo clinico de gravedad " + gravedad);
        }
        return ticket;
    }

    public Ticket escalarTicket(int numero, String motivo)
    {
        Ticket ticket = buscarTicket(numero);
        if (!flujo.puedeCambiarEstado(ticket.getEstado(), "Escalado"))
        {
            throw new IllegalStateException("No se puede escalar un ticket en estado '" + ticket.getEstado() + "'.");
        }
        Veterinario senior = null;
        for (Veterinario v : veterinarios)
        {
            boolean compatible = v.getEspecialidad().equalsIgnoreCase(ticket.getTipoConsulta())
                || v.getEspecialidad().equalsIgnoreCase("General");
            if (compatible && (senior == null || v.getNivelExperiencia() > senior.getNivelExperiencia()))
            {
                senior = v;
            }
        }
        ticket.escalar(motivo, senior);
        return ticket;
    }

    public Ticket resolverTicket(int numero, String diagnostico, String tratamiento)
    {
        Ticket ticket = buscarTicket(numero);
        if (!flujo.puedeCambiarEstado(ticket.getEstado(), "Resuelto"))
        {
            throw new IllegalStateException("No se puede resolver un ticket en estado '" + ticket.getEstado() + "'.");
        }
        ticket.resolver(diagnostico, tratamiento);
        return ticket;
    }

    public Ticket cerrarTicket(int numero)
    {
        Ticket ticket = buscarTicket(numero);
        if (!flujo.puedeCambiarEstado(ticket.getEstado(), "Cerrado"))
        {
            throw new IllegalStateException("No se puede cerrar un ticket en estado '" + ticket.getEstado() + "'.");
        }
        ticket.cerrar();
        return ticket;
    }

    private Veterinario asignarVeterinarioAutomatico(String tipoConsulta)
    {
        Veterinario elegido = null;
        for (Veterinario v : veterinarios)
        {
            if (v.getEspecialidad().equalsIgnoreCase(tipoConsulta) && v.estaDisponible())
            {
                if (elegido == null || v.getCargaActual() < elegido.getCargaActual())
                {
                    elegido = v;
                }
            }
        }
        if (elegido == null)
        {
            for (Veterinario v : veterinarios)
            {
                if (v.getEspecialidad().equalsIgnoreCase("General") && v.estaDisponible())
                {
                    if (elegido == null || v.getCargaActual() < elegido.getCargaActual())
                    {
                        elegido = v;
                    }
                }
            }
        }
        return elegido;
    }

    public Ticket buscarTicket(int numero)
    {
        for (Ticket t : tickets)
        {
            if (t.getNumero() == numero)
            {
                return t;
            }
        }
        throw new NoSuchElementException("No existe un ticket con numero " + numero + ".");
    }

    public void mostrarTickets()
    {
        if (tickets.isEmpty())
        {
            System.out.println("No hay tickets registrados.");
            return;
        }
        for (Ticket t : tickets)
        {
            t.mostrarResumen();
        }
    }

    public void mostrarFlujo()
    {
        flujo.mostrarFlujo();
    }

    public void generarMetricas()
    {
        System.out.println("Total de tickets: " + tickets.size());

        String[] estados = { "Abierto", "Asignado", "Escalado", "Resuelto", "Cerrado" };
        System.out.println("Por estado:");
        for (String estado : estados)
        {
            int contador = 0;
            for (Ticket t : tickets)
            {
                if (t.getEstado().equals(estado))
                {
                    contador++;
                }
            }
            System.out.println("  " + estado + ": " + contador);
        }

        String[] prioridades = { "Baja", "Media", "Alta", "Emergencia" };
        System.out.println("Por prioridad:");
        for (String prioridad : prioridades)
        {
            int contador = 0;
            for (Ticket t : tickets)
            {
                if (t.getPrioridad().equalsIgnoreCase(prioridad))
                {
                    contador++;
                }
            }
            System.out.println("  " + prioridad + ": " + contador);
        }

        int escalados = 0;
        for (Ticket t : tickets)
        {
            if (t.isEscalado())
            {
                escalados++;
            }
        }
        System.out.println("Tickets escalados: " + escalados);

        System.out.println("Carga por veterinario:");
        for (Veterinario v : veterinarios)
        {
            System.out.println("  " + v.getNombre() + " (" + v.getEspecialidad() + "): " + v.getCargaActual() + "/" + v.getCapacidadMaxima());
        }
    }

    public ArrayList<Veterinario> getVeterinarios()
    {
        return veterinarios;
    }

    public ArrayList<Cliente> getClientes()
    {
        return clientes;
    }

    public ArrayList<Mascota> getMascotas()
    {
        return mascotas;
    }

    public ArrayList<Ticket> getTickets()
    {
        return tickets;
    }
}
